tools/autograd/derivatives.yaml

```diff
@@ -968,8 +968,8 @@
 - name: sgn(Tensor self) -> Tensor
   self: sgn_backward(result, grad, self)
 
-- name: sin(Tensor self) -> Tensor
-  self: grad * self.cos().conj()
+- name: sin(Tensor self, *, ScalarType? dtype=None) -> Tensor
+  self: (grad * self.cos().conj()).to(self.dtype())
```

Comment:
This is really interesting. We'll need to call @albanD to determine the appropriate behavior for gradient propagation.

@albanD, I made a note for us this Friday

Reply:
Some reference:
https://github.com/pytorch/pytorch/blob/ad7d208ba5f2c5614679a7999918b75ae74530e9/tools/autograd/derivatives.yaml#L372-L376

Reply:
From the discussion on Friday, I think this the expected behavior as long as no non-continuous dtype is involved. Indeed, this is exactly what the `.to()` formula does as well.

For non-contiguous dtypes, we should not compute gradients and the autograd will raise an error if you try to output such a Tensor and attach gradient to it. So you might want to add special tests here to make sure it has the behavior you want.